File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
    //     unordered_map<Node*,Node*> nn;
    //     Node* h=head;
    //     while (h){
    //         nn[h]=new Node (h->val);
    //         h=h->next;
    //     }
    //     h=head;
    //     while (h){
    //         nn[h]->next=nn[h->next];
    //         nn[h]->random=nn[h->random];
    //         h=h->next;
    //     }
    //     return nn[head];
    // }

            Node* temp=head;
            while(temp){
                Node* copy= new Node (temp->val);
                copy->next=temp->next;
                temp->next=copy;
                temp=copy->next;
            }
            temp=head;
            while(temp){
                if (temp->random) temp->next->random=temp->random->next;
                temp=temp->next->next;
            }
            Node* dum=new Node(-1);
            Node* h=dum;
            temp=head;
            while(temp){
                h->next=temp->next;
                h=h->next;

                temp->next=h->next;
                temp=temp->next;
            }
            return dum->next;
    }


};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head: return None 
        nn={}
        h=head
        while h: 
            nn[h]=Node(h.val)
            h=h.next
        h=head
        while h:
            nn[h].next=nn.get(h.next)
            nn[h].random=nn.get(h.random)
            h=h.next
        return nn[head]

        